### PR TITLE
Mass matrix quad

### PR DESCRIPTION
**Pull request by inducer**
This adds two new mass matrix functions, plus a slew of more-or-less related cleanups. Best read commit-by-commit.

### CODE REVIEW

File: modepy/matrices.py [hunk 1]
```diff
@@ -229,17 +241,31 @@ def differentiation_matrices(basis, grad_basis, nodes, from_nodes=None):
...
-    if isinstance(grad_vdms, tuple):
-        return tuple(
-                np.asarray(la.solve(vdm.T, gv.T).T, order="C")
-                for gv in grad_vdms)
-    else:
-        return np.asarray(
-                la.solve(vdm.T, grad_vdms.T).T,
-                order="C")
+    if len(basis) != from_nodes.shape[-1]:
+        raise ValueError("basis is not unisolvent, cannot interpolate")
```
Comment (alexfikl) [inline modepy/matrices.py:245]: Put this check before the Vandermonde matrices are computed?

File: modepy/matrices.py [hunk 2]
```diff
@@ -313,8 +339,68 @@ def mass_matrix(
...
+        ) -> np.ndarray:
+    r"""Using the *quadrature*, provide a matrix :math:`M` that
+    satisfies:
+
+    .. math::
+
+        \displaystyle (M \boldsymbol u)_i = \sum_j w_j \phi_i(r_j) u_j,
+
+    where :math:`\phi_i` are the (volume) Lagrange basis functions obtained from
+    *test_functions* at *nodes*, :math:`w_i` and :math:`r_i` are the
```
Comment (alexfikl) [inline modepy/matrices.py:379]:
```suggestion
    *test_functions* at *nodes*, :math:`w_j` and :math:`r_j` are the
```
? To match the indices in the formula.

File: modepy/matrices.py [hunk 3]
```diff
@@ -369,34 +459,62 @@ def nodal_mass_matrix_for_face(face: Face, face_quad: Quadrature,
...
     satisfies:
 
     .. math::
 
         \displaystyle (M_f \boldsymbol u)_i = \sum_j w_j \phi_i(r_j) u_j,
 
-    where :math:`\phi_i` are the (volume) Lagrange basis functions obtained from
-    *test_functions* at *volume_nodes*, and :math:`w_i` and :math:`r_i` are the
-    weights and nodes from *face_quad*.
+    where :math:`\phi_i` are the *test_functions*, :math:`w_i` and :math:`r_i` are
```
Comment (alexfikl) [inline modepy/matrices.py:476]:
```suggestion
    where :math:`\phi_i` are the *test_functions*, :math:`w_j` and :math:`r_j` are
```
?

**Review by alexfikl — APPROVED** [review]
Left a couple of very nitpicky comments, but otherwise this looks nice and straightforward to me :grin:

**Comment by inducer** [comment]
Thanks for taking a look!

> Also, regarding that lumping in grudge. Is it possible with this interface? If I understand correctly, it would just mean passing a Gauss-Legendre quadrature or a Lobatto-G-L to integrate in nodal/modal_quad_mass_matrix?

I'm planning to add a separate function that returns just the diagonal, after appropriate checking. But this seemed like a nice, self-contained set of changes, so I figured I'd do that separately. I'm also still puzzling over the interface.

File: modepy/matrices.py (hunk 2 above)
Comment (inducer) [reply modepy/matrices.py:379]: Done

File: modepy/matrices.py (hunk 3 above)
Comment (inducer) [reply modepy/matrices.py]: Done

**Pull request merged**